### PR TITLE
sql: support implicit record arg and return types in UDFs

### LINKED ISSUE #86393: sql: udf to support implicit record argument/return type
**Describe the solution you'd like**
Right now, implicit record return/argument types in UDF are resolved but unimplemented error is returned because execution doesn't support that yet. We need to add support in execution before we can remove the unimplemented error.

### PR DESCRIPTION
**Pull request by mgartner**
#### opt: wrap last statement in a UDF with LIMIT 1

A UDF always returns a single row from the last statement in the UDF
(unless it is a set returning UDF, but we do not currently support
those). So, we can wrap the last statement in a `LIMIT 1` expression
which may allow additional optimizations to trigger.

Release justification: This is a minor change to a new feature.

Release note: None

#### opt: add implicit record types to opt test catalog

Release justification: This is a test-only change.

Release note: None

#### sql: support implicit record arg and return types in UDFs

Implicit record types are now supported as argument types and return
types of user-defined functions. For example:

    CREATE TABLE t (a INT PRIMARY KEY, b INT);
    CREATE FUNCTION ret_t() RETURNS t LANGUAGE SQL AS $$
      SELECT * FROM t
    $$;
    CREATE FUNCTION t_a(some_t t) RETURNS INT LANGUAGE SQL AS $$
      SELECT (some_t).a
    $$;

Fixes #86393

Release justification: This fixes a limitation of a newly added feature.

Release note: None

#### sql: propagate UDF errors correctly

This commit fixes a bug where an error that occurred during evaluation
of a UDF was not correctly propagated.

Release justification: This fixes a critical bug with a new feature.

Release note: None

#### sql: cast UDF result to function return type

The last statement of a UDF may have a resulting type, `R`, that does
not match the function's given return type, `F`. This is allowed if a
cast from `R` to `F` is valid in an implicit or assignment context.

This commit adds an assignment cast operator to the output column of the
final statement in the function body, if necessary.

Release justification: This fixes a minor bug with a new feature.

Release Note: None

### CODE REVIEW

File: pkg/sql/logictest/testdata/logic_test/udf [hunk 1]
```diff
@@ -1539,3 +1533,153 @@ l1  l2  i1  i2  s1  s2  v1  v2
...
+statement ok
+CREATE FUNCTION imp_const_tup() RETURNS imp LANGUAGE SQL AS $$
+  SELECT (11, 22, 'b')
+$$
+
+# TODO(mgartner): The type of imp_const_tup() should be imp, not record.
+query TBT
+SELECT imp_const_tup(), (11,22,'b')::imp = imp_const_tup(), pg_typeof(imp_const_tup())
+----
+(11,22,b)  true  record
```
Comment (chengxiong-ruan) [inline pkg/sql/logictest/testdata/logic_test/udf:1553]: heh, looks like it returns `imp` instead of `record` in `fakedist-disk_test` test config.

Comment (mgartner) [reply pkg/sql/logictest/testdata/logic_test/udf:1553]: Weird.. Ok I'll have to debug why it's `record` in the local config.

**Review by DrewKimball — APPROVED** [review]
Nice work! After you've figured out whatever's going on with the `record` typing, :lgtm:

> Reviewed 5 of 5 files at r1, 2 of 2 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-:-NAv-6XH-NYp_cn6DNP2:bqrkbch)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan, @mgartner, and @msirek)

___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 1560 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAuw9tbED2Ep1Yb7P9Z:-NAuw9tbED2Ep1Yb7P9_:b-n1wsxd) ([raw file](https://github.com/cockroachdb/cockroach/blob/97dab7f0ea4867766e1876be932de6b77620a3b4/pkg/sql/logictest/testdata/logic_test/udf#L1560)):*
> ```
> $$
> 
> # TODO(mgartner): The type of imp_const_tup() should be imp, not record.
> ```

[nit] The type is alread `imp` on this one.
___
*[`pkg/sql/opt/optbuilder/scalar.go` line 667 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAutSZlCbHDVmIskVrF:-NAutSZlCbHDVmIskVrG:bevuab) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4c58b8f5df0a54fecf189ffa83e472bebc64c1a/pkg/sql/opt/optbuilder/scalar.go#L667)):*
> ```Go
> 		// rows after the first can simply be ignored. The limit could be
> 		// beneficial because it could allow additional optimization.
> 		if lastStmt := i == len(stmts)-1; lastStmt {
> ```

[very nit] I feel like this is more confusing than just using `i == len(stmts)-1`, or pulling the assignment out to the previous line.

**Review by msirek — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @chengxiong-ruan and @mgartner)

___
*[`pkg/sql/opt/testutils/testcat/types.go` line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAvULzN-JZ-mlIg0OX-:-NAvULzOAkrneA9PgDyj:b-mjp8vt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6696974b2e7b386f3c44d6a18b47b929203c658f/pkg/sql/opt/testutils/testcat/types.go#L71)):*
> ```Go
> 			contents := make([]*types.T, 0, tab.ColumnCount())
> 			labels := make([]string, 0, tab.ColumnCount())
> 			for i, n := 0, tab.ColumnCount(); i < n; i++ {
> ```

Since we're only building a tuple in the multicolumn case, maybe we shouldn't be trying to execute this code when `tab.ColumnCount()` is 1, but break out and return the error below.

**Review by mgartner — COMMENTED** [review]
To fix the type issues I had to implement assignment-casting of return values. I also caught a bug where errors were not being correctly propagated. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-:-NAzfTNt3rD6N8M67Yo1:b-kalm6p)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @DrewKimball, and @msirek)

___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 1560 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAuw9tbED2Ep1Yb7P9Z:-NAzfS8LE23WKiB_emRQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/97dab7f0ea4867766e1876be932de6b77620a3b4/pkg/sql/logictest/testdata/logic_test/udf#L1560)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] The type is alread `imp` on this one.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/optbuilder/scalar.go` line 667 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAutSZlCbHDVmIskVrF:-NAzAVBU5P_YOExtZcD-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c4c58b8f5df0a54fecf189ffa83e472bebc64c1a/pkg/sql/opt/optbuilder/scalar.go#L667)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[very nit] I feel like this is more confusing than just using `i == len(stmts)-1`, or pulling the assignment out to the previous line.
</blockquote></details>

Done.
___
*[`pkg/sql/opt/testutils/testcat/types.go` line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAvULzN-JZ-mlIg0OX-:-NAz9cT-DcRwRAeOnUaW:bwbb9lq) ([raw file](https://github.com/cockroachdb/cockroach/blob/6696974b2e7b386f3c44d6a18b47b929203c658f/pkg/sql/opt/testutils/testcat/types.go#L71)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Since we're only building a tuple in the multicolumn case, maybe we shouldn't be trying to execute this code when `tab.ColumnCount()` is 1, but break out and return the error below.
</blockquote></details>

Implicit record types can have a single column - if the table that they represent has a single column. For example:

```
defaultdb> CREATE TABLE t (a INT);
CREATE TABLE

defaultdb> SELECT row(1)::t, (row(1)::t).a;
  row | a
------+----
  (1) | 1
(1 row)
```

I've added a test to make sure the test catalog can handle this case correctly.

**Review by DrewKimball — COMMENTED** [review]
> Reviewed 6 of 6 files at r4, 2 of 2 files at r5, 4 of 4 files at r6, 2 of 2 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @DrewKimball, @mgartner, and @msirek)

___
*[`pkg/sql/routine.go` line 58 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NB-559zCvE_S6d65VkW:-NB-559zCvE_S6d65VkX:bpca3u2) ([raw file](https://github.com/cockroachdb/cockroach/blob/96594185020053b3824eac72f587707e023b8be1/pkg/sql/routine.go#L58)):*
> ```Go
> 			_ = p.Txn().ConfigureStepping(ctx, prevSteppingMode)
> 			seqErr := txn.SetReadSeqNum(prevSeqNum)
> 			// Propagate the original error, if there is one.
> ```

Should this logic even be executed if there was already an error?
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 1710 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NB-6YacBqwSJSZjW8D4:-NB-6YacBqwSJSZjW8D5:b6f515j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d634e94ce1c88422c6ad028f372fb0b72ba090d/pkg/sql/logictest/testdata/logic_test/udf#L1710)):*
> ```
> SELECT stoc('a'), pg_typeof(stoc('a'))
> ----
> a  text
> ```

I'd think the type would be `character` - is there some bug here (maybe not related to udfs) or is this expected behavior?
```
postgres=# select pg_typeof('b'::CHAR);
 pg_typeof
-----------
 character
(1 row)
```
vs
```
root@localhost:26257/defaultdb> select pg_typeof('b'::CHAR);

  pg_typeof
-------------
  text
(1 row)
```

**Review by msirek — COMMENTED** [review]
> Reviewed 3 of 5 files at r1, 6 of 6 files at r4, 2 of 2 files at r5, 4 of 4 files at r6, 2 of 2 files at r7, 4 of 4 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan, @DrewKimball, and @mgartner)

___
*[`pkg/sql/opt/testutils/testcat/types.go` line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAvULzN-JZ-mlIg0OX-:-NB-X6YZARWacoMSwvZC:bou8m4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/6696974b2e7b386f3c44d6a18b47b929203c658f/pkg/sql/opt/testutils/testcat/types.go#L71)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Implicit record types can have a single column - if the table that they represent has a single column. For example:

```
defaultdb> CREATE TABLE t (a INT);
CREATE TABLE

defaultdb> SELECT row(1)::t, (row(1)::t).a;
  row | a
------+----
  (1) | 1
(1 row)
```

I've added a test to make sure the test catalog can handle this case correctly.
</blockquote></details>

OK, thanks.  Do we want to qualify the name of the object type, including the schema (if it differs from the current schema) as in postgres?

```
CREATE TABLE t (a INT);
CREATE SCHEMA mws;
CREATE TABLE mws.t (a char(10));
SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);

crdb:
SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);
  pg_typeof | pg_typeof
------------+------------
  t         | t
(1 row)

postgres:
postgres=# SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);
 pg_typeof | pg_typeof 
-----------+-----------
 t         | mws.t

postgres=# set schema 'mws';
SET
postgres=# SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);
 pg_typeof | pg_typeof 
-----------+-----------
 t         | t
```
I'm not sure if we'd want to qualify database name too.  Postgres doesn't support such references:
```
implicit=# SELECT pg_typeof(row(1)::postgres.mws.t);
ERROR:  cross-database references are not implemented: postgres.mws.t

(1 row)
```

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan and @DrewKimball)

___
*[`pkg/sql/routine.go` line 58 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NB-559zCvE_S6d65VkW:-NBSLeh99SsZ2hWZHMA5:bx4jy95) ([raw file](https://github.com/cockroachdb/cockroach/blob/96594185020053b3824eac72f587707e023b8be1/pkg/sql/routine.go#L58)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Should this logic even be executed if there was already an error?
</blockquote></details>

That's a good point. If the UDF errors, it should abort the txn, so I don't think we need to do this. I've updated it.
___
*[`pkg/sql/logictest/testdata/logic_test/udf` line 1710 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NB-6YacBqwSJSZjW8D4:-NBSKnpK1ynpgG7M9IVR:b-xi1anx) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d634e94ce1c88422c6ad028f372fb0b72ba090d/pkg/sql/logictest/testdata/logic_test/udf#L1710)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

I'd think the type would be `character` - is there some bug here (maybe not related to udfs) or is this expected behavior?
```
postgres=# select pg_typeof('b'::CHAR);
 pg_typeof
-----------
 character
(1 row)
```
vs
```
root@localhost:26257/defaultdb> select pg_typeof('b'::CHAR);

  pg_typeof
-------------
  text
(1 row)
```
</blockquote></details>

It's a known issue tracked by https://github.com/cockroachdb/cockroach/issues/58252.
___
*[`pkg/sql/opt/testutils/testcat/types.go` line 71 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/87275#-NAvULzN-JZ-mlIg0OX-:-NBSKw4oAzauBmr72ZJz:bmm0vgj) ([raw file](https://github.com/cockroachdb/cockroach/blob/6696974b2e7b386f3c44d6a18b47b929203c658f/pkg/sql/opt/testutils/testcat/types.go#L71)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

OK, thanks.  Do we want to qualify the name of the object type, including the schema (if it differs from the current schema) as in postgres?

```
CREATE TABLE t (a INT);
CREATE SCHEMA mws;
CREATE TABLE mws.t (a char(10));
SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);

crdb:
SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);
  pg_typeof | pg_typeof
------------+------------
  t         | t
(1 row)

postgres:
postgres=# SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);
 pg_typeof | pg_typeof 
-----------+-----------
 t         | mws.t

postgres=# set schema 'mws';
SET
postgres=# SELECT pg_typeof(row(1)::t), pg_typeof(row(1)::mws.t);
 pg_typeof | pg_typeof 
-----------+-----------
 t         | t
```
I'm not sure if we'd want to qualify database name too.  Postgres doesn't support such references:
```
implicit=# SELECT pg_typeof(row(1)::postgres.mws.t);
ERROR:  cross-database references are not implemented: postgres.mws.t

(1 row)
```
</blockquote></details>

We could qualify the names and make this more robust, but we generally only add the bare minimum functionality to the test catalog to keep it simple. Since I don't need to test qualified names in the optbuilder tests I added, I left it out.

**Review by DrewKimball — APPROVED** [review]
> Reviewed 8 of 8 files at r9, 2 of 2 files at r10, 4 of 4 files at r11, 2 of 2 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/87275)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @chengxiong-ruan and @DrewKimball)

**Review by chengxiong-ruan — APPROVED** [review]

**Comment by mgartner** [comment]
TFTRs!

bors r+

**Comment by rafiss** [comment]
blathers backport 22.2

**Pull request merged**